File: MTCS.py
import numpy as np
import math

import torch
import torch.nn.functional as F


class MonteCarloTree:
    def __init__(self, value_network, device, exploration_factor=5):
        self.value_network = value_network
        self.root = None
        self.device = device
        self.exploration_factor = exploration_factor

    def simulate(self, game):
        node = self.root
        while not node.is_leaf():
            action, node = node.select_child(self.exploration_factor)
            game.make_move(action)

        if game.is_game_over():
            winner = game.check_winner()
            if winner == game.current_player:
                value = 1
            elif winner == game.get_other_player():
                value = -1
            else:
                value, prior_prob = self.evaluate_state(game.get_state())
        else:
            value, prior_prob = self.evaluate_state(game.get_state())
            node.expand(game, prior_prob)

        self.backpropagate(node, -value)

    def search(self, game, node, num_simulations):
        self.root = node

        for _ in range(num_simulations):
            self.simulate(game.copy())

    def evaluate_state(self, state):
        state_tensor = torch.from_numpy(state).unsqueeze(0).float().to(self.device)  # 将状态转换为张量
        value, policy = self.value_network(state_tensor)  # 使用策略价值网络评估状态

        value = value[0][0].cpu().item()  # 将值转换为标量
        prior_prob = np.exp(policy.data.cpu().numpy().flatten())
        return value, prior_prob

    def backpropagate(self, node, value):
        while node is not None:
            node.update(value)
            node = node.parent
            value = -value

    def get_action_probabilities(self, game, temperature=1.0):
        node = self.root
        action_visits = [(action, child.visits) for action, child in node.children.items()]
        actions, visits = zip(*action_visits)
        visits_tensor = torch.tensor(visits, dtype=torch.float)
        action_probs = F.softmax(1.0 / temperature * torch.log(visits_tensor + 1e-10), dim=0).tolist()

        probs = [0] * game.board_size * game.board_size
        for i in range(len(actions)):
            probs[game.get_action_index(actions[i])] = action_probs[i]

        return np.array(range(game.board_size * game.board_size)), np.array(probs)

    def apply_temperature(self, action_probabilities, temperature):
        if temperature == 1:
            return action_probabilities
        action_probabilities = np.power(action_probabilities, 1 / temperature)
        action_probabilities /= np.sum(action_probabilities)
        return action_probabilities


class Node:
    def __init__(self, parent=None):
        self.parent = parent
        self.children = {}
        self.visits = 0
        self.value_sum = 0
        self.prior_prob = 0

    def is_leaf(self):
        return len(self.children) == 0

    def select_child(self, exploration_factor):
        total_visits = self.visits
        ucb_values = [
            ((child.value_sum / child.visits) if child.visits > 0 else 0) +
            exploration_factor * child.prior_prob * math.sqrt(total_visits) /
            (1 + child.visits) for i, child in
            enumerate(self.children.values())]
        selected_action = None
        max_ucb = float('-inf')
        for i, action in enumerate(self.children):
            if ucb_values[i] > max_ucb:
                max_ucb = ucb_values[i]
                selected_action = action
        return selected_action, self.children[selected_action]

    def expand(self, game, prior_probs):
        actions = game.get_valid_actions()
        for i, action in enumerate(actions):
            child = Node(parent=self)
            child.prior_prob = prior_probs[game.get_action_index(action)]
            self.children[action] = child

    def update(self, value):
        self.visits += 1
        self.value_sum += value

File: MTCS.cpp
#include <iostream>
#include <cmath>
#include <unordered_map>
#include <vector>
#include "Game.cpp"

class MonteCarloTree
{
public:
    MonteCarloTree(torch::nn::Module network, torch::Device device, float exploration_factor = 5)
        : network(network), root(nullptr), device(device), exploration_factor(exploration_factor) {}

    void simulate(Game game)
    {
        if (game.isGameOver())
            return;

        Node *node = root;
        while (!node->isLeaf())
        {
            std::pair<int, Node *> result = node->selectChild(exploration_factor);
            int action = result.first;
            node = result.second;
            Point pointAction = game.getPointFormIndex(action);
            game.makeMove(pointAction);
        }

        std::pair<float, std::vector<float>> result = evaluate_state(game.getState());
        float value = result.first;
        std::vector<float> priorProb = result.second;
        if (game.checkWin(game.lastAction.x, game.lastAction.y, game.getOtherPlayer()))
        {
            value = -1;
        }
        else
        {
            node->expand(game, priorProb);
        }

        backpropagate(node, -value);
    }

    void search(Game game, Node *node, int num_simulations)
    {
        root = node;

        for (int i = 0; i < num_simulations; i++)
            simulate(game);
    }

    std::pair<float, std::vector<float>> evaluate_state(torch::Tensor state)
    {
        torch::Tensor state_tensor = state.to(device).clone();
        std::pair<torch::Tensor, torch::Tensor> result = network->forward(state_tensor);
        torch::Tensor value = result.first;
        torch::Tensor policy = result.second;

        float valueFloat = value[0][0].item<float>();
        std::vector<float> prior_prob = torch::exp(policy).cpu().data().flatten().tolist<float>();
        return std::make_pair(valueFloat, prior_prob);
    }

    void backpropagate(Node *node, float value)
    {
        while (node != nullptr)
        {
            node->update(value);
            node = node->parent;
            value = -value;
        }
    }

    std::pair<std::vector<int>, std::vector<float>> get_action_probabilities(Game game, float temperature = 1.0)
    {
        Node *node = root;
        std::vector<std::pair<Action, int>> action_visits;
        for (auto &item : node->children)
        {
            action_visits.push_back(std::make_pair(item.first, item.second->visits));
        }

        std::vector<Action> actions;
        std::vector<int> visits;
        for (auto &item : action_visits)
        {
            actions.push_back(item.first);
            visits.push_back(item.second);
        }

        torch::Tensor visits_tensor = torch::from_blob(visits.data(), {visits.size()}).clone();
        std::vector<float> action_probs = torch::softmax(1.0 / temperature * torch::log(visits_tensor + 1e-10), 0).tolist<float>();

        std::vector<float> probs(game.board_size * game.board_size, 0);
        for (int i = 0; i < actions.size(); i++)
        {
            if (visits[i] == 1)
                action_probs[i] = 0;
            probs[game.get_action_index(actions[i])] = action_probs[i];
        }

        std::vector<int> range(game.board_size * game.board_size);
        std::iota(range.begin(), range.end(), 0);
        return std::make_pair(range, probs);
    }

    std::vector<float> apply_temperature(std::vector<float> action_probabilities, float temperature)
    {
        if (temperature == 1)
            return action_probabilities;
        for (float &prob : action_probabilities)
        {
            prob = std::pow(prob, 1 / temperature);
        }
        float sum = std::accumulate(action_probabilities.begin(), action_probabilities.end(), 0.0f);
        for (float &prob : action_probabilities)
        {
            prob /= sum;
        }
        return action_probabilities;
    }

private:
    torch::nn::Module network;
    Node *root;
    torch::Device device;
    float exploration_factor;
};

class Node
{
public:
    Node *parent;
    int visits;
    double value_sum;
    double prior_prob;
    std::unordered_map<int, Node *> children;

    Node(Node *parent = nullptr) : parent(parent), visits(0), value_sum(0), prior_prob(0) {}

    bool isLeaf()
    {
        return children.empty();
    }

    std::pair<int, Node *> selectChild(double exploration_factor)
    {
        int total_visits = visits;
        std::vector<double> ucb_values;
        for (const auto &child : children)
        {
            double ucb_value = (child.second->value_sum / child.second->visits) +
                               exploration_factor * child.second->prior_prob *
                                   std::sqrt(total_visits) / (1 + child.second->visits);
            ucb_values.push_back(ucb_value);
        }
        int selected_action = -1;
        double max_ucb = std::numeric_limits<double>::lowest();
        for (int i = 0; i < ucb_values.size(); i++)
        {
            if (ucb_values[i] > max_ucb)
            {
                max_ucb = ucb_values[i];
                selected_action = i;
            }
        }
        return std::make_pair(selected_action, children[selected_action]);
    }

    void expand(Game &game, const std::vector<float> &prior_probs)
    {
        std::vector<Point> actions = game.getEmptyPoints();
        for (int i = 0; i < actions.size(); i++)
        {
            Node *child = new Node(this);
            int actionIndex = game.getActionIndex(actions[i]);
            child->prior_prob = prior_probs[actionIndex];
            children[actionIndex] = child;
        }
    }

    void update(double value)
    {
        visits++;
        value_sum += value;
    }
};
